src/app/clusters/descriptor/descriptor.cpp

```diff
@@ -0,0 +1,172 @@
+    {
+        cluster = emberAfGetNthCluster(endpoint, clusterIndex, server);
+        status  = writeAttribute(endpoint, attributeId, (uint8_t *) &cluster->clusterId, clusterIndex);
+        VerifyOrReturnError(status == EMBER_ZCL_STATUS_SUCCESS, status);
+    }
+
+    return writeAttribute(endpoint, attributeId, (uint8_t *) &clusterCount);
+}
+
+EmberAfStatus writeServerAttribute(uint8_t endpoint)
+{
+    return writeClientServerAttribute(endpoint, true);
+}
+
+EmberAfStatus writeClientAttribute(uint8_t endpoint)
+{
+    return writeClientServerAttribute(endpoint, false);
+}
+
+EmberAfStatus writeDeviceAttribute(uint8_t endpoint)
+{
+    EmberAfStatus status    = EMBER_ZCL_STATUS_SUCCESS;
+    AttributeId attributeId = ZCL_DEVICE_ATTRIBUTE_ID;
+
+    uint16_t deviceTypeCount = 0;
+    uint8_t startIndex       = endpoint;
+    uint8_t stopIndex        = (endpoint == 0 ? emberAfEndpointCount() : static_cast<uint8_t>(endpoint + 1));
+
+    EmberAfDeviceType deviceType;
+    for (uint8_t endpointIndex = startIndex; endpointIndex < stopIndex; endpointIndex++)
+    {
+        DeviceId deviceId = emberAfDeviceIdFromIndex(endpointIndex);
+        uint16_t revision = emberAfDeviceVersionFromIndex(endpointIndex);
+
+        bool duplicate = false;
+        for (uint8_t entryIndex = 0; entryIndex < (endpointIndex - endpoint); entryIndex++)
+        {
+            status = readAttribute(endpoint, attributeId, (uint8_t *) &deviceType, endpointIndex - endpoint);
+            VerifyOrReturnError(status == EMBER_ZCL_STATUS_SUCCESS, status);
+
+            if (deviceType.deviceTypeId == deviceId)
+            {
+                duplicate = true;
+                break;
+            }
+        }
+
+        if (duplicate == false)
+        {
+            deviceType.deviceTypeId = deviceId;
+            deviceType.revision     = revision;
+
+            status = writeAttribute(endpoint, attributeId, (uint8_t *) &deviceType, endpointIndex - endpoint);
+            VerifyOrReturnError(status == EMBER_ZCL_STATUS_SUCCESS, status);
+            deviceTypeCount++;
+        }
+    }
+
+    return writeAttribute(endpoint, attributeId, (uint8_t *) &deviceTypeCount);
+}
+
+EmberAfStatus writePartsAttribute(uint8_t endpoint)
+{
+    EmberAfStatus status    = EMBER_ZCL_STATUS_SUCCESS;
+    AttributeId attributeId = ZCL_PARTS_ATTRIBUTE_ID;
+
+    uint16_t partsCount = 0;
+
+    for (uint8_t endpointIndex = 0; endpointIndex < emberAfEndpointCount(); endpointIndex++)
+    {
+        EndpointId endpointId = emberAfEndpointFromIndex(endpointIndex);
+
+        // The Descriptor Cluster instance on Endpoint 0 (zero) SHALL have a Parts attribute that lists all endpoint instances on a
+        // node instance.
+        // All local endpoints required by the device type(s) in the Device attribute SHALL be included in the Parts list.
+        if (endpoint == 0x00 || (emberAfDeviceIdFromIndex(endpoint) == emberAfDeviceIdFromIndex(endpointIndex)))
```

Comment:
I'm not quite following that second condition.  You can have multiple same-type devices on the node, and devices don't span endpoints...

This hierarchical endpoint relationship will need to be expressed in the app config (i.e. the .zap file) and get spit out by codegen somewhere; I don't think our current generated code representes it.  Please file a followup issue to do that?